### PR TITLE
JS module loader: switch from Bower + RequireJS to jspm + SystemJS

### PR DESCRIPTION
**Pull request by OliverJAsh**
This brings our entire codebase to a working, production ready(?) state.

Whilst we test that our application still runs fine with SystemJS and we didn't miss anything whilst migrating, we are still building and serving the RequireJS bundle (from the same source code). The SystemJS bundle will be delivered to a small proportion of users until we are comfortable it is all working OK. So, for now, all the RequireJS configuration remains alongside SystemJS.

To do before merge:
- [x] Create A/B test to distribute traffic
- [x] Thoroughly QA JS rich areas of the site
- [x] Review `TODO` comments — some of these might be small things worth fixing/tidying up now
- [x] Prefetch bundles
- [x] Review bundles — are we bundling everything correctly, or are dependencies still being loaded from source/`jspm_packages`?
- [x] Check all jspm dependency versions are the same as Bower

/cc @guybedford @theefer

### CODE REVIEW

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 1]
```diff
@@ -335,6 +323,118 @@
...
+        @if(!play.Play.isDev()) {
+            @Html(Static.js.systemJsBundleConfig)
+        }
+
+        System.paths['ophan/ng']  = '@{Configuration.javascript.config("ophanJsUrl")}.js';
+        System.paths['googletag'] = '@{Configuration.javascript.config("googletagJsUrl")}';
+
+        @Html(Static.js.systemJsNormalize)
+
+        window.require = System.amdRequire.bind(System);
```
Comment (OliverJAsh) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: The fix for this is in System 0.16.8 and 0.16.9 https://github.com/systemjs/systemjs/releases/tag/0.16.9

File: facia-tool/test/public/utils/collections-loader.js [hunk 2]
```diff
@@ -6,9 +6,9 @@ define([
     'test/fixtures/articles',
     'test/fixtures/some-collections',
     'utils/layout-from-url',
-    'views/collections.scala.html!text',
-    'views/templates/vertical_layout.scala.html!text',
-    'widgets/collection.html!text',
+    'text!views/collections.scala.html',
+    'text!views/templates/vertical_layout.scala.html',
+    'text!widgets/collection.html',
```
Comment (OliverJAsh) [inline facia-tool/test/public/utils/collections-loader.js]: My find+replace mistakingly changed the format for facia-tool. Revert

File: package.json [hunk 3]
```diff
@@ -36,7 +36,8 @@
     "handlebars": "2.0.0",
     "jit-grunt": "0.8.0",
     "jsonfile": "2.0.0",
-    "jspm": "0.14.0",
+    "jspm": "0.15.0",
+    "jspm-bower-endpoint": "^0.3.2",
```
Comment (OliverJAsh) [inline package.json]: Dependencies brought down from bower: `socketio` and `videojspersistvolume`. The former can be fetched from npm (`socket.io-client` IIRC), the later can be fetched from GitHub. After discussions with @guybedford, it turns out the GitHub endpoint/registry does everything the Bower one does, including reading the Bower/package manifest for `main`.

File: package.json [hunk 4]
```diff
@@ -64,19 +67,40 @@
...
+      "videojspersistvolume": "bower:videojs-persistvolume@~0.1.0",
+      "videojsplaylist": "github:tim-peterson/videojs-playlist@master"
+    },
+    "devDependencies": {
+      "babel": "npm:babel@^4.7.16",
+      "babel-runtime": "npm:babel-runtime@^4.7.16"
     }
+  },
+  "dependencies": {
+    "mkdirp": "^0.5.0"
```
Comment (OliverJAsh) [inline package.json]: Move to `devDeps`.

File: static/src/javascripts/projects/common/utils/inlineSvg.js [hunk 5]
```diff
@@ -20,7 +20,7 @@ define([
                 dirs = path.split("/"),
                 fileName = dirs.pop();
 
-            text.get(req.toUrl(dirs.join("/") + "/" + imageType + "/" + fileName + ".svg"), function (svg) {
+            text.get(req.toUrl(dirs.join("/") + "/" + (imageType ? imageType + "/" : "") + fileName + ".svg"), function (svg) {
```
Comment (OliverJAsh) [inline static/src/javascripts/projects/common/utils/inlineSvg.js]: The tests were failing without this change. Does the plugin still need this `imageType` magic?

**Comment by OliverJAsh** [comment]
Started https://github.com/guardian/frontend/issues/9007 to discuss next steps after merging.

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 6]
```diff
@@ -335,6 +323,118 @@
...
+        @Html(Static.js.es6ModuleLoader)
+        @Html(Static.js.systemJs)
+        @Html(Static.js.systemJsAppConfig)
+
+        @if(!play.Play.isDev()) {
+            @Html(Static.js.systemJsBundleConfig)
+        }
+
+        System.paths['ophan/ng']  = '@{Configuration.javascript.config("ophanJsUrl")}.js';
+        System.paths['googletag'] = '@{Configuration.javascript.config("googletagJsUrl")}';
```
Comment (OliverJAsh) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Curious why Ophan needed `.js` but Google Tag didn't…

Comment (rich-nguyen) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: just because the configs are just string properties on file, and one has .js, the other doesn't

File: common/app/CorsErrorHandler.scala [hunk 7]
```diff
@@ -9,11 +9,16 @@ trait CorsErrorHandler extends GlobalSettings with Results with common.Execution
...
+      super.onError(request, ex)
+    } else {
+      val headers = request.headers
+      val vary = headers.get("Vary").fold(defaultVaryFields)(v => (v :: varyFields).mkString(","))
 
-    Future.successful{
-      Cors(InternalServerError.withHeaders("Vary" -> vary))(request)
+      Future.successful {
+        Cors(InternalServerError.withHeaders("Vary" -> vary))(request)
+      }
```
Comment (phamann) [inline common/app/CorsErrorHandler.scala]: These changes – although v.useful – probably belong in a separate PR.

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 8]
```diff
@@ -335,6 +323,118 @@
...
+
+@bootSystemJs() = {
+
+    <script>
+        @Html(Static.js.es6ModuleLoader)
+        @Html(Static.js.systemJs)
+        @Html(Static.js.systemJsAppConfig)
+
+        @if(!play.Play.isDev()) {
+            @Html(Static.js.systemJsBundleConfig)
```
Comment (phamann) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Have we tested this is PROD mode yet?

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 9]
```diff
@@ -335,6 +323,118 @@
...
+
+        System.paths['ophan/ng']  = '@{Configuration.javascript.config("ophanJsUrl")}.js';
+        System.paths['googletag'] = '@{Configuration.javascript.config("googletagJsUrl")}';
+
+        @Html(Static.js.systemJsNormalize)
+
+        window.require = System.amdRequire.bind(System);
+        window.define = System.amdDefine.bind(System);
+
+        Promise.all(
```
Comment (phamann) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: What is this magic `Promise` object you speak of?? We don't have they polyfill loaded for this at this point and therefore will throw on older devices...

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 10]
```diff
@@ -335,6 +323,118 @@
...
+                        }
+
+                        return @conf.Switches.DiagnosticsLogging.isSwitchedOn &&
+                            Math.random() < 0.2 &&
+                            @{!play.Play.isDev()}; @* don't actually notify sentry in dev mode*@
+                    }
+                }
+            ).install();
+
+            Promise.all([
```
Comment (phamann) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: As above. Sadly, it's not safe to use `Promise` here.

File: grunt-configs/uglify.js [hunk 11]
```diff
@@ -1,17 +1,44 @@
...
+                expand: true,
+                cwd: 'static/src',
+                src: ['systemjs-normalize.js',
+                      'systemjs-config.js',
+                      'systemjs-bundle-config.js'],
+                dest: 'common/conf/assets'
+            }],
+            options:   {
+                compress:{
+                    evaluate: false
```
Comment (phamann) [inline grunt-configs/uglify.js]: Maybe leave a comment above as to why we needed to do this.

File: static/src/systemjs-normalize.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
+// We use system.normalize to convert RequireJS module IDs to SystemJS module
+// IDs. This is mainly necessary because the plugin format is reverted in
+// SystemJS (suffix rather than prefix).
+
+// Used in JavaScriptLaterSteps.scala.html and bundle.js
+// IIFE is the only way to share this piece of code.
+// If only we had a module loader to load the module loader.
```
Comment (phamann) [inline static/src/systemjs-normalize.js]: :100:

**Comment by phamann** [comment]
> Review TODO comments — some of these might be small things worth fixing/tidying up now

Agree, most of these i'd like cleaning up before we merge..

**Comment by phamann** [comment]
:+1: Can't see anything blindingly obvious in here. 

This is *A.M.A.Z.I.N.G* work. Can't wait to see this working in production.

**Comment by theefer** [comment]
Awesome work indeed.

File: common/app/CorsErrorHandler.scala (hunk 7 above)
Comment (rich-nguyen) [inline common/app/CorsErrorHandler.scala]: In #9021

File: common/app/views/fragments/javaScriptLaterSteps.scala.html (hunk 8 above)
Comment (rich-nguyen) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: I've been testing bundles in Prod today, have submitted fixes to issues regarding bundle arithmetic

File: common/app/views/fragments/javaScriptLaterSteps.scala.html (hunk 9 above)
Comment (rich-nguyen) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Courtesy of es6-module-loader, as discussed

File: common/app/views/fragments/javaScriptLaterSteps.scala.html (hunk 8 above)
Comment (OliverJAsh) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Out of interest can you link to these?

On Mon, 27 Apr 2015 at 16:44 Richard Nguyen <notifications@github.com>
wrote:

> In common/app/views/fragments/javaScriptLaterSteps.scala.html
> <https://github.com/guardian/frontend/pull/9005#discussion_r29158649>:
>
> > @@ -335,6 +323,118 @@
> >                  });
> >
> >              });
> > +    </script>
> > +}
> > +
> > +@bootSystemJs() = {
> > +
> > +    <script>
> > +        @Html(Static.js.
>
> es6ModuleLoader)
> > +        @Html(Static.js.systemJs)
> > +        @Html(Static.js.systemJsAppConfig)
> > +
> > +        @if(!play.Play.isDev()) {
> > +            @Html(Static.js.systemJsBundleConfig)
>
> I've been testing bundles in Prod today, have submitted fixes to issues
> regarding bundle arithmetic
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/9005/files#r29158649>.
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

Comment (rich-nguyen) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: All in the commit list for today

**Comment by rich-nguyen** [comment]
There are some version mismatches, but I think they are all minor positive upgrades. Everything green, going to ship tomorrow after some testing and feedback from QA

File: README.md [hunk 13]
```diff
@@ -115,10 +115,14 @@ sudo npm -g install grunt-cli
 Ubuntu/Mac:
 ```
 sudo npm -g install jspm
-jspm endpoint config github
+jspm registry config github
```
Comment (OliverJAsh) [inline README.md]: Should we add this to `setup.sh`?

Comment (rich-nguyen) [inline README.md]: The user needs to provide the github application token.

Comment (OliverJAsh) [inline README.md]: OK, might be worth sending out an email when this goes out to let devs know they will need to do the step below.

**Comment by rich-nguyen** [comment]
Ze test matrix.

For each loader, and for each entry, I shall be checking front, article, live blog, nav, video, adverts, analytics, discussion. Anything else I've missed?

New JSPM loader 
- [] Firefox
- [] Explorer
- [] Safari
- [] Chrome
- [] IOS
- [] Android

Current Curl loader
- [] Firefox
- [] Explorer
- [] Safari
- [] Chrome
- [] IOS
- [] Android

**Comment by ScottPainterGNM** [comment]
@rich-nguyen Looks like a good list, lets catch up after the stand up on this and which are the best operating systems for testing some of these browsers on.

**Comment by ScottPainterGNM** [comment]
Happy from a QA perspective apart from the aforementioned comments about making sure that network fronts are shorter for IPad1/IE8 as they are in prod. Not sure this would impact or not but just that it appeared when on a local host of this branch.

**Comment by rich-nguyen** [comment]
thanks @ScottPainterGNM
